source/code/plugins/oms_diag.rb

```diff
@@ -0,0 +1,135 @@
+module OMS
+  class Diag
+
+    # Defaults for diagnostic
+    DEFAULT_IPNAME = "Diagnostics"
+    DEFAULT_TAG    = "diag.oms"
+
+    # Mandatory property keys
+    DI_KEY_LOGMESSAGE = 'LogData'
+    DI_KEY_IPNAME     = 'IPName'
+    DI_KEY_TYPE       = 'type'
+    DI_KEY_TIME       = 'time'
+    DI_KEY_AGENTGUID  = 'sourceHealthServiceId'
+
+    # Constants for dataitem values
+    DI_TYPE_XML   = 'System.PropertyBagData'
+    DI_TYPE_JSON  = 'JsonData'
+
+    # Record keys
+    RECORD_DATAITEMS  = 'DataItems'
+    RECORD_IPNAME     = 'IPName'
+    RECORD_MGID       = 'ManagementGroupId'
+
+    # Record constant values
+    RECORD_MGID_VALUE = '{00000000-0000-0000-0000-000000000002}'
+
+    class << self
+
+      # Method to be used by INPUT and FILTER plugins for logging
+
+      # Method:
+      # LogDiag(logMessage, tag=DEFAULT_TAG)
+      #
+      # Description:
+      # This is to be utilized for logging to the diagnostic
+      # channel.
+      #
+      # Parameters:
+      # @logMessage[mandatory]: The log message string to be logged
+      # @tag[optional]: The tag with which to emit the diagnostic log. The
+      # default value would be DEFAULT_TAG
+      # @ipname[optional]: IPName can be optionally provided to depict a
+      # customized one other than the DEFAULT_IPNAME in diagnostic event
+      # @properties[optional]: Hash corresponding to key value pairs that
+      # would be added as part of this data item.
+      #
+      # NOTE: Certain mandatory properties to the dataitem are added by default
+      def LogDiag(logMessage, tag=DEFAULT_TAG, ipname=DEFAULT_IPNAME, properties=nil)
+        # Process default values for tag and ipname if they are passed as nil
+        tag ||= DEFAULT_TAG
+        ipname ||= DEFAULT_IPNAME
+
+        dataitem = Hash.new
+
+        # Adding parameterized properties
+        dataitem.merge!(properties) if properties.is_a?(Hash)
+
+        # Adding mandatory properties
+        dataitem[DI_KEY_LOGMESSAGE] = logMessage
+        dataitem[DI_KEY_IPNAME]     = ipname
+        dataitem[DI_KEY_TIME]       = Time.now.utc.strftime("%Y-%m-%dT%H:%M:%S.%6NZ")
```

Comment:
The Time format present in diagnostic logs is different from the one given out in oms_common.

- Format expected in the diagnostic logs (as per existing) is "%Y-%m-%dT%H:%M:%S.%6NZ"

- Format provided by format_time_str of oms_common is %FT%H:%M:%S.%3NZ